enhancements/microshift/microshift-updateability-ostree.md

```diff
@@ -0,0 +1,613 @@
+- If persisted version is missing and version of binary is 4.14, assume 4.13 was persisted and allow storage migration.
+
+### Data migration
+
+Data migration shall include:
+- Storage version - update of Kubernetes objects by reading Resource in older version and writing newer version
+  - Leveraging existing [Kube Storage Version Migrator](https://github.com/openshift/kubernetes-kube-storage-version-migrator) and [its Operator](https://github.com/openshift/cluster-kube-storage-version-migrator-operator).
+- etcd schema - comparing v2 and v3, most likely if/when etcd v4 is released
+  - etcd project documents how to migrate from v2 to v3, but we'll also ask OpenShift etcd team for guidance.
+
+### Open Questions [optional]
+
+- Do we want to persist ostree commit ref in the metadata?
+  - Can we make some part of the process safer, more robust, by comparing previous and current commit ref?
+  - Are we good with having previous boot MicroShift version persisted?
+- If green/red info is not persisted, and it's not a first boot (/var/lib/microshift exists), what should we do?
+  - Compare versions (metadata vs binary)
+    - Same - start cluster
+    - Different - refuse to start? Try to migrate (but create backup before)?
+
+- Should we use greenboot's green/red scripts to persist action for next boot?
+  - They have no information about what failed, so if it wasn't MicroShift, then we do unnecessary restore (possibly losing some data)
+
+- Do we want to persist the history of boots or defer to checking `boot_counter`?
+  - Red + no `boot_counter` => first boot after rollback
+  - Do we want to read kernel cmdline?
+    - Seems to be internal detail of implementation
+
+- How to we avoid endless reboots in any case? Are they possible?
+  - Maybe we can persist short history of previous boots/actions on boot?
+    - How that would protect us? If we refuse the start, greenboot checks will fail, so maybe we need override it?
+
+- How should `microshift pre-run` be executed?
+  - `microshift.service` - `ExecStartPre`
+    - No need to add new systemd service files.
+    - It will run on each `systemctl restart microshift` which is not desirable (will it run when systemd restarts MicroShift?)
+  - `microshift-pre-run.service`
+    - Running on boot, just once, before `microshift.service`
+    - Not repeated on MicroShift restart
+    - New service file
+
+### Test Plan
+
+#### Unit tests
+
+Aiming to write as much as possible in Go, we should strive for maximum testability: 
+- Separate code paths for planning (e.g. should it do a backup or restore?) 
+  and acting (actually perform backup) - e.g. interface with two methods Plan(), Act()
+  - This will allow testing decisions and actions separately
+  - This will allow easy implementation of --dry-run describing what would happen
+- Due to many interactions with filesystem, a virtual filesystem or 
```

Comment:
What does "virtual filesystem" mean here?

Reply:
I had this library in mind: https://github.com/spf13/afero

Reply:
I'll reword it in the enhancement to be more clear